### PR TITLE
kernel_cmdline: Fix parsing/equivalence of "outside" quoted args

### PR DESCRIPTION
**Pull request by jeckersb**
Primary motivation here is that these two should be equivalent:

foo="quoted value"
"foo=quoted value"

This also adds tests for a few more oddball cases that weren't covered
before but clarify the expected kernel behavior.

Closes: #1737
Signed-off-by: John Eckersberg <jeckersb@redhat.com>

### CODE REVIEW

File: crates/kernel_cmdline/src/bytes.rs [hunk 1]
```diff
@@ -549,9 +557,22 @@ mod tests {
     fn test_parameter_pathological() {
         // valid things that certified insane people would do
 
-        // quotes don't get removed from keys
```
Comment (jeckersb) [inline crates/kernel_cmdline/src/bytes.rs:552]: This was just incorrect before.

File: crates/kernel_cmdline/src/bytes.rs [hunk 2]
```diff
@@ -421,27 +421,31 @@ impl<'a> Parameter<'a> {
...
-                // *Only* the first and last double quotes are stripped
-                value = {
-                    let v = value.strip_prefix(b"\"").unwrap_or(value);
-                    v.strip_suffix(b"\"").unwrap_or(v)
-                };
+                // If there is a quote after the equals, skip it.  If
+                // there was a closing quote at the end of the value,
+                // we would have already removed it in
+                // `dequoted_input` above
+                value = value.strip_prefix(b"\"").unwrap_or(value);
```
[the remark this answers is not in the document]
Comment (jeckersb) [reply crates/kernel_cmdline/src/bytes.rs:448]: I don't believe you but it's time to go home so I will look at it tomorrow :laughing:

[the remark this answers is not in the document]
Comment (jeckersb) [reply crates/kernel_cmdline/src/bytes.rs:448]: Ok not quite tomorrow but there's a passing test for that case...

```rust
    #[test]
    fn test_parameter_quoted() {
        let p = param("foo=\"quoted value\"");
        assert_eq!(p.value, Some(b"quoted value".as_slice()));
```

Plus this whole bit just doesn't make any sense:

> For an argument like `foo="bar"`, the closing quote is at the end of the value but not at the end of the whole parameter, so it isn't removed by the initial dequoted_input step.

:-1:

File: crates/kernel_cmdline/src/bytes.rs [hunk 3]
```diff
@@ -421,27 +421,31 @@ impl<'a> Parameter<'a> {
     /// This is an internal method that assumes the input has already been
     /// split into a single parameter (e.g., by CmdlineIterBytes).
     fn parse_internal(input: &'a [u8]) -> Option<Self> {
-        let equals = input.iter().position(|b| *b == b'=');
+        // *Only* the first and last double quotes are stripped
+        let dequoted_input = input.strip_prefix(b"\"").unwrap_or(input);
+        let dequoted_input = dequoted_input.strip_suffix(b"\"").unwrap_or(dequoted_input);
```
Comment (cgwalters) [inline crates/kernel_cmdline/src/bytes.rs:426]: What happens if there's only one quote at the start? Shouldn't we *not* strip the trailing one in that case? What does the Linux kernel do in this corner case?

File: crates/kernel_cmdline/src/utf8.rs [hunk 4]
```diff
@@ -426,6 +426,16 @@ mod tests {
     fn test_parameter_quoted() {
         let p = param("foo=\"quoted value\"");
         assert_eq!(p.value(), Some("quoted value"));
+
+        let p = param("foo=\"unclosed quotes");
+        assert_eq!(p.value(), Some("unclosed quotes"));
```
Comment (cgwalters) [inline crates/kernel_cmdline/src/utf8.rs:431]: I guess this is an answer to my question above, but is it the correct answer?

**Review by cgwalters — APPROVED** [review]

File: crates/kernel_cmdline/src/utf8.rs (hunk 4 above)
Comment (jeckersb) [reply crates/kernel_cmdline/src/utf8.rs:431]: Yeah that's this line - https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/lib/cmdline.c#n256.  We see '=', replace it with '\0', advance one, if it's '"' we advance again past it, effectively stripping it.
